File: ingest/build_archive.py
#!/usr/bin/env python3
"""Build the normalized vintage-correct archive from data/raw/ (see docs/schema.md).

Reads only the local cache — no network. Emits data/archive/{forecasts,outcomes}
as parquet + csv mirror, append-only semantics: the build is deterministic from
the raw cache, so "appending" happens by refreshing the cache, not by editing
outputs. Every row carries provenance (source_file, retrieved_at).
"""

import io
import json
import re
import zipfile
from datetime import date
from pathlib import Path

import pandas as pd

ROOT = Path(__file__).resolve().parent.parent
RAW = ROOT / "data" / "raw"
OUT = ROOT / "data" / "archive"
E = dict(engine="calamine")

MANIFEST = json.loads((RAW / "manifest.json").read_text())


def retrieved(fname: str) -> str:
    return MANIFEST[fname]["retrieved_at"]


def saar_to_qq(s: pd.Series) -> pd.Series:
    return ((1 + s / 100) ** 0.25 - 1) * 100


def quarter_label(d) -> str:
    d = pd.Timestamp(d)
    return f"{d.year}Q{d.quarter}"


# --- GDPNow: forecasts (TrackingDeepArchives + TrackingArchives) + outcomes (TrackRecord) ---

def gdpnow():
    f = "gdpnow_tracking.xlsx"
    frames = []
    for sheet in ["TrackingDeepArchives", "TrackingArchives"]:
        df = pd.read_excel(RAW / f, sheet_name=sheet, **E)
        df = df.rename(columns=lambda c: str(c).strip())
        keep = df[["Forecast Date", "Quarter being forecasted", "GDP Nowcast"]].dropna()
        keep = keep.assign(sheet=sheet)
        frames.append(keep)
    fc = pd.concat(frames, ignore_index=True)
    # overlap between the two tabs: keep the regular archive's row
    fc = fc.sort_values("sheet").drop_duplicates(
        subset=["Forecast Date", "Quarter being forecasted"], keep="last"
    )
    fc = pd.DataFrame({
        "source": "gdpnow",
        "region": "US",
        "variable": "rgdp_growth",
        "target_period": fc["Quarter being forecasted"].map(quarter_label),
        "forecast_date": pd.to_datetime(fc["Forecast Date"]).dt.date,
        "output_type": "point",
        "output_id": None,
        "value_native": fc["GDP Nowcast"].astype(float),
        "unit_native": "saar_pct",
        "declared_target": "advance",  # GDPNow TrackRecord scores itself vs BEA advance
        "retrieved_at": retrieved(f),
        "source_file": f,
    })

    tr = pd.read_excel(RAW / f, sheet_name="TrackRecord", skiprows=1, header=None, **E)
    tr = tr.iloc[:, [0, 2, 3]].dropna()
    oc = pd.DataFrame({
        "region": "US",
        "variable": "rgdp_growth",
        "target_period": tr[0].map(quarter_label),
        "release_label": "advance",
        "published_on": pd.to_datetime(tr[3]).dt.date,
        "value_native": tr[2].astype(float),
        "unit_native": "saar_pct",
        "source": "gdpnow_trackrecord(BEA)",
        "retrieved_at": retrieved(f),
        "source_file": f,
    })
    return fc, oc


def gdpnow_current():
    """In-flight quarter from CurrentQtrEvolution (not yet in the archive tabs):
    [Date, Major Releases, GDP*] column triplets; quarter parsed from the
    'Initial GDPNow YY:Qq forecast' label."""
    f = "gdpnow_tracking.xlsx"
    df = pd.read_excel(RAW / f, sheet_name="CurrentQtrEvolution", header=None, **E)
    m = None
    for cell in df.astype(str).to_numpy().ravel():
        mm = re.search(r"Initial GDPNow (\d\d):Q(\d) forecast", cell)
        if mm:
            m = mm
            break
    if not m:
        return pd.DataFrame()
    tp = f"20{m.group(1)}Q{m.group(2)}"
    rows = []
    for c in range(0, df.shape[1] - 2, 3):
        dates = pd.to_datetime(df[c], errors="coerce")
        vals = pd.to_numeric(df[c + 2], errors="coerce")
        ok = dates.notna() & vals.notna()
        rows.append(pd.DataFrame({"forecast_date": dates[ok].dt.date, "value": vals[ok]}))
    cur = pd.concat(rows, ignore_index=True).drop_duplicates(subset=["forecast_date"])
    return pd.DataFrame({
        "source": "gdpnow",
        "region": "US",
        "variable": "rgdp_growth",
        "target_period": tp,
        "forecast_date": cur["forecast_date"],
        "output_type": "point",
        "output_id": None,
        "value_native": cur["value"].astype(float),
        "unit_native": "saar_pct",
        "declared_target": "advance",
        "retrieved_at": retrieved(f),
        "source_file": f"{f}:CurrentQtrEvolution",
    })


# --- NY Fed Staff Nowcast: Forecasts By Quarter grid (2023- relaunch file) ---

def nyfed():
    f = "nyfed_staff_nowcast.xlsx"
    raw = pd.read_excel(RAW / f, sheet_name="Forecasts By Quarter", header=None, **E)
    hdr_idx = raw.index[raw[0].astype(str).str.strip() == "Forecast Date"][0]
    hdr = raw.iloc[hdr_idx]
    df = raw.iloc[hdr_idx + 1:].copy()
    df.columns = hdr
    df = df[pd.to_datetime(df["Forecast Date"], errors="coerce").notna()]
    qcols = [c for c in df.columns if re.fullmatch(r"\d{4}Q[1-4]", str(c))]
    long = df.melt(id_vars=["Forecast Date"], value_vars=qcols,
                   var_name="target_period", value_name="value").dropna(subset=["value"])
    return pd.DataFrame({
        "source": "nyfed",
        "region": "US",
        "variable": "rgdp_growth",
        "target_period": long["target_period"].astype(str),
        "forecast_date": pd.to_datetime(long["Forecast Date"]).dt.date,
        "output_type": "point",
        "output_id": None,
        "value_native": long["value"].astype(float),
        "unit_native": "saar_pct",
        "declared_target": "unspecified",
        "retrieved_at": retrieved(f),
        "source_file": f,
    })


# --- Philadelphia Fed SPF: mean current-quarter growth (drgdp2) per survey round ---

PHILLY_ROUND_MONTH = {1: 2, 2: 5, 3: 8, 4: 11}  # survey published mid Feb/May/Aug/Nov


def spf_philly():
    f = "spf_philly_meangrowth.xlsx"
    df = pd.read_excel(RAW / f, sheet_name="RGDP", **E).dropna(subset=["drgdp2"])
    # drgdp2 = mean forecast for the survey's own quarter (the SPF "nowcast"), SAAR %.
    # Exact deadline dates aren't in this file: forecast_date is approximated as the
    # 15th of the round's middle month (documented in ingest/README.md).
    return pd.DataFrame({
        "source": "spf_philly",
        "region": "US",
        "variable": "rgdp_growth",
        "target_period": df["YEAR"].astype(int).astype(str) + "Q" + df["QUARTER"].astype(int).astype(str),
        "forecast_date": [date(int(y), PHILLY_ROUND_MONTH[int(q)], 15)
                          for y, q in zip(df["YEAR"], df["QUARTER"])],
        "output_type": "point",
        "output_id": None,
        "value_native": df["drgdp2"].astype(float),
        "unit_native": "saar_pct",
        "declared_target": "unspecified",
        "retrieved_at": retrieved(f),
        "source_file": f,
    })


# --- ECB SPF: mean GDP point forecast per round & target (year-on-year, kept native) ---

GDP_HDR = "GROWTH EXPECTATIONS"


def spf_ecb():
    f = "spf_ecb_individual.zip"
    rows = []
    with zipfile.ZipFile(RAW / f) as z:
        for name in sorted(z.namelist()):
            m = re.fullmatch(r"(\d{4})Q([1-4])\.csv", name)
            if not m:
                continue
            year, q = int(m.group(1)), int(m.group(2))
            txt = z.read(name).decode("utf-8", errors="replace")
            df = pd.read_csv(io.StringIO(txt), header=None, dtype=str, engine="python",
                             on_bad_lines="skip")
            col0 = df[0].fillna("")
            # section headers = prose rows (letters, not TARGET_PERIOD); data rows start
            # with a target period like "2026", "2026Q4", "2027Mar"
            is_hdr = col0.str.contains(r"[A-Za-z]") & (col0 != "TARGET_PERIOD") & \
                ~col0.str.fullmatch(r"\d{4}(Q[1-4]|[A-Z][a-z]{2})?")
            hdr_idx = list(col0[is_hdr].index)
            gdp_start = [i for i in hdr_idx if GDP_HDR in col0[i]]
            if not gdp_start:
                continue
            start = gdp_start[0]
            later = [i for i in hdr_idx if i > start]
            end = later[0] if later else len(df)
            sec = df.iloc[start + 1:end]
            hdr_row = sec[sec[0] == "TARGET_PERIOD"]
            if hdr_row.empty:
                continue
            hdr = hdr_row.iloc[0]
            point_col = hdr[hdr == "POINT"].index[0]
            body = sec.loc[hdr_row.index[0] + 1:]
            body = body[body[0].notna() & (body[0] != "")]
            vals = pd.to_numeric(body[point_col], errors="coerce")
            grp = pd.DataFrame({"target": body[0], "point": vals}).dropna()
            for target, mean in grp.groupby("target")["point"].mean().items():
                rows.append({
                    "source": "spf_ecb",
                    "region": "EA",
                    "variable": "rgdp_growth",
                    "target_period": str(target),
                    # ECB SPF rounds close in the first month of the quarter (approx: 15th)
                    "forecast_date": date(year, (q - 1) * 3 + 1, 15),
                    "output_type": "point",
                    "output_id": None,
                    "value_native": round(float(mean), 4),
                    "unit_native": "yoy_pct",
                    "declared_target": "unspecified",
                    "retrieved_at": retrieved(f),
                    "source_file": f"{f}:{name}",
                })
    return pd.DataFrame(rows)


def main():
    OUT.mkdir(parents=True, exist_ok=True)
    gd_fc, gd_oc = gdpnow()
    gd_cur = gdpnow_current()
    if len(gd_cur):
        gd_fc = pd.concat([gd_fc, gd_cur], ignore_index=True).drop_duplicates(
            subset=["target_period", "forecast_date"], keep="first")
    forecasts = pd.concat([gd_fc, nyfed(), spf_philly(), spf_ecb()], ignore_index=True)
    forecasts["value_qq"] = forecasts["value_native"].where(
        forecasts["unit_native"] == "saar_pct").pipe(saar_to_qq).round(4)
    # yoy rows are not identifiable to q/q -> value_qq stays NaN by construction

    outcomes = gd_oc.copy()
    outcomes["value_qq"] = saar_to_qq(outcomes["value_native"]).round(4)

    col_order_fc = ["source", "region", "variable", "target_period", "forecast_date",
                    "output_type", "output_id", "value_native", "unit_native", "value_qq",
                    "declared_target", "retrieved_at", "source_file"]
    col_order_oc = ["region", "variable", "target_period", "release_label", "published_on",
                    "value_native", "unit_native", "value_qq", "source", "retrieved_at",
                    "source_file"]
    forecasts = forecasts[col_order_fc].sort_values(
        ["source", "target_period", "forecast_date"]).reset_index(drop=True)
    outcomes = outcomes[col_order_oc].sort_values(["target_period"]).reset_index(drop=True)

    for name, df in [("forecasts", forecasts), ("outcomes", outcomes)]:
        df.to_parquet(OUT / f"{name}.parquet", index=False)
        df.to_csv(OUT / f"{name}.csv", index=False)
        print(f"{name}: {len(df):,} rows -> data/archive/{name}.parquet (+.csv)")
    print("\nforecast rows per source:")
    print(forecasts.groupby("source").agg(rows=("value_native", "size"),
                                          first=("forecast_date", "min"),
                                          last=("forecast_date", "max")).to_string())


if __name__ == "__main__":
    main()

File: ingest/build_archive.R
#!/usr/bin/env Rscript
# Build the normalized vintage-correct archive from data/raw/ (see docs/schema.md).
#
# Reads only the local cache — no network. Emits data/archive/{forecasts,outcomes}
# as parquet + csv mirror, append-only semantics: the build is deterministic from
# the raw cache, so "appending" happens by refreshing the cache, not by editing
# outputs. Every row carries provenance (source_file, retrieved_at).

suppressMessages(library(data.table))

ROOT <- normalizePath(file.path(dirname(sub("--file=", "", grep("--file=", commandArgs(FALSE), value = TRUE))), ".."))
RAW <- file.path(ROOT, "data", "raw")
OUT <- file.path(ROOT, "data", "archive")

MANIFEST <- jsonlite::read_json(file.path(RAW, "manifest.json"))
retrieved <- function(fname) MANIFEST[[fname]]$retrieved_at

saar_to_qq <- function(s) ((1 + s / 100)^0.25 - 1) * 100

quarter_label <- function(d) {
  d <- as.Date(d)
  sprintf("%dQ%d", as.integer(format(d, "%Y")), (as.integer(format(d, "%m")) - 1) %/% 3 + 1)
}

# --- GDPNow: forecasts (TrackingDeepArchives + TrackingArchives) + outcomes (TrackRecord) ---

gdpnow <- function() {
  f <- "gdpnow_tracking.xlsx"
  path <- file.path(RAW, f)
  frames <- lapply(c("TrackingDeepArchives", "TrackingArchives"), function(sheet) {
    df <- suppressWarnings(readxl::read_excel(path, sheet = sheet))
    names(df) <- trimws(names(df))
    keep <- as.data.table(df[, c("Forecast Date", "Quarter being forecasted", "GDP Nowcast")])
    setnames(keep, c("fdate", "qdate", "value"))
    keep <- keep[!is.na(fdate) & !is.na(qdate) & !is.na(value)]
    keep[, sheet := sheet]
    keep
  })
  fc <- rbindlist(frames)
  # overlap between the two tabs: same behavior as the validated Python build —
  # rows sorted by sheet name, last kept per (forecast date, quarter)
  setorder(fc, sheet)
  fc <- fc[!duplicated(fc, by = c("fdate", "qdate"), fromLast = TRUE)]
  forecasts <- data.table(
    source = "gdpnow", region = "US", variable = "rgdp_growth",
    target_period = quarter_label(fc$qdate),
    forecast_date = as.Date(fc$fdate),
    output_type = "point", output_id = NA_character_,
    value_native = as.numeric(fc$value), unit_native = "saar_pct",
    declared_target = "advance",  # GDPNow TrackRecord scores itself vs BEA advance
    retrieved_at = retrieved(f), source_file = f
  )

  tr <- suppressWarnings(readxl::read_excel(path, sheet = "TrackRecord", skip = 1, col_names = FALSE))
  tr <- as.data.table(tr[, c(1, 3, 4)])
  setnames(tr, c("qdate", "value", "pub"))
  tr <- tr[!is.na(qdate) & !is.na(value) & !is.na(pub)]
  outcomes <- data.table(
    region = "US", variable = "rgdp_growth",
    target_period = quarter_label(tr$qdate),
    release_label = "advance",
    published_on = as.Date(tr$pub),
    value_native = as.numeric(tr$value), unit_native = "saar_pct",
    source = "gdpnow_trackrecord(BEA)",
    retrieved_at = retrieved(f), source_file = f
  )
  list(fc = forecasts, oc = outcomes)
}

gdpnow_current <- function() {
  # In-flight quarter from CurrentQtrEvolution (not yet in the archive tabs):
  # [Date, Major Releases, GDP*] column triplets; quarter parsed from the
  # 'Initial GDPNow YY:Qq forecast' label.
  f <- "gdpnow_tracking.xlsx"
  df <- suppressWarnings(readxl::read_excel(file.path(RAW, f), sheet = "CurrentQtrEvolution",
                                            col_names = FALSE, col_types = "text"))
  cells <- unlist(df, use.names = FALSE)
  m <- regmatches(cells, regexec("Initial GDPNow (\\d\\d):Q(\\d) forecast", cells))
  m <- Filter(function(x) length(x) == 3, m)
  if (!length(m)) return(data.table())
  tp <- sprintf("20%sQ%s", m[[1]][2], m[[1]][3])
  rows <- rbindlist(lapply(seq(1, ncol(df) - 2, by = 3), function(c0) {
    d <- suppressWarnings(as.Date(as.numeric(df[[c0]]), origin = "1899-12-30"))
    v <- suppressWarnings(as.numeric(df[[c0 + 2]]))
    ok <- !is.na(d) & !is.na(v)
    data.table(forecast_date = d[ok], value = v[ok])
  }))
  rows <- rows[!duplicated(forecast_date)]
  data.table(
    source = "gdpnow", region = "US", variable = "rgdp_growth",
    target_period = tp, forecast_date = rows$forecast_date,
    output_type = "point", output_id = NA_character_,
    value_native = rows$value, unit_native = "saar_pct",
    declared_target = "advance",
    retrieved_at = retrieved(f), source_file = paste0(f, ":CurrentQtrEvolution")
  )
}

# --- NY Fed Staff Nowcast: Forecasts By Quarter grid (2023- relaunch file) ---

nyfed <- function() {
  f <- "nyfed_staff_nowcast.xlsx"
  raw <- suppressWarnings(readxl::read_excel(file.path(RAW, f), sheet = "Forecasts By Quarter",
                                             col_names = FALSE, col_types = "text"))
  hdr_row <- which(trimws(raw[[1]]) == "Forecast Date")[1]
  hdr <- unlist(raw[hdr_row, ], use.names = FALSE)
  body <- raw[-seq_len(hdr_row), , drop = FALSE]
  names(body) <- ifelse(is.na(hdr), paste0("x", seq_along(hdr)), hdr)
  dt <- as.data.table(body)
  dt[, fdate := as.Date(as.numeric(`Forecast Date`), origin = "1899-12-30")]
  dt <- dt[!is.na(fdate)]
  qcols <- grep("^\\d{4}Q[1-4]$", names(dt), value = TRUE)
  long <- melt(dt, id.vars = "fdate", measure.vars = qcols,
               variable.name = "target_period", value.name = "value", variable.factor = FALSE)
  long[, value := suppressWarnings(as.numeric(value))]
  long <- long[!is.na(value)]
  data.table(
    source = "nyfed", region = "US", variable = "rgdp_growth",
    target_period = long$target_period, forecast_date = long$fdate,
    output_type = "point", output_id = NA_character_,
    value_native = long$value, unit_native = "saar_pct",
    declared_target = "unspecified",
    retrieved_at = retrieved(f), source_file = f
  )
}

# --- Philadelphia Fed SPF: mean current-quarter growth (drgdp2) per survey round ---

PHILLY_ROUND_MONTH <- c(2, 5, 8, 11)  # survey published mid Feb/May/Aug/Nov

spf_philly <- function() {
  f <- "spf_philly_meangrowth.xlsx"
  df <- as.data.table(readxl::read_excel(file.path(RAW, f), sheet = "RGDP"))
  df <- df[!is.na(drgdp2)]
  # drgdp2 = mean forecast for the survey's own quarter (the SPF "nowcast"), SAAR %.
  # Exact deadline dates aren't in this file: forecast_date is approximated as the
  # 15th of the round's middle month (documented in ingest/README.md).
  data.table(
    source = "spf_philly", region = "US", variable = "rgdp_growth",
    target_period = sprintf("%dQ%d", as.integer(df$YEAR), as.integer(df$QUARTER)),
    forecast_date = as.Date(sprintf("%d-%02d-15", as.integer(df$YEAR),
                                    PHILLY_ROUND_MONTH[as.integer(df$QUARTER)])),
    output_type = "point", output_id = NA_character_,
    value_native = as.numeric(df$drgdp2), unit_native = "saar_pct",
    declared_target = "unspecified",
    retrieved_at = retrieved(f), source_file = f
  )
}

# --- ECB SPF: mean GDP point forecast per round & target (year-on-year, kept native) ---

GDP_HDR <- "GROWTH EXPECTATIONS"

spf_ecb <- function() {
  f <- "spf_ecb_individual.zip"
  zpath <- file.path(RAW, f)
  entries <- sort(grep("^\\d{4}Q[1-4]\\.csv$", utils::unzip(zpath, list = TRUE)$Name, value = TRUE))
  out <- lapply(entries, function(name) {
    yr <- as.integer(substr(name, 1, 4)); qn <- as.integer(substr(name, 6, 6))
    lines <- readLines(unz(zpath, name), warn = FALSE)
    fields <- strsplit(lines, ",", fixed = TRUE)
    col0 <- vapply(fields, function(x) if (length(x)) x[1] else "", "")
    # section headers = prose rows (letters, not TARGET_PERIOD); data rows start
    # with a target period like "2026", "2026Q4", "2027Mar"
    is_hdr <- grepl("[A-Za-z]", col0) & col0 != "TARGET_PERIOD" &
      !grepl("^\\d{4}(Q[1-4]|[A-Z][a-z]{2})?$", col0)
    hdr_idx <- which(is_hdr)
    gdp_start <- hdr_idx[grepl(GDP_HDR, col0[hdr_idx], fixed = TRUE)]
    if (!length(gdp_start)) return(NULL)
    start <- gdp_start[1]
    later <- hdr_idx[hdr_idx > start]
    end <- if (length(later)) later[1] else length(lines) + 1L
    sec <- seq(start + 1L, end - 1L)
    trow <- sec[col0[sec] == "TARGET_PERIOD"][1]
    if (is.na(trow)) return(NULL)
    point_col <- which(fields[[trow]] == "POINT")[1]
    body <- sec[sec > trow & col0[sec] != ""]
    vals <- suppressWarnings(as.numeric(vapply(fields[body], function(x)
      if (length(x) >= point_col) x[point_col] else NA_character_, "")))
    grp <- data.table(target = col0[body], point = vals)[!is.na(point)]
    if (!nrow(grp)) return(NULL)
    agg <- grp[, .(value = round(mean(point), 4)), keyby = target]
    data.table(
      source = "spf_ecb", region = "EA", variable = "rgdp_growth",
      target_period = agg$target,
      # ECB SPF rounds close in the first month of the quarter (approx: 15th)
      forecast_date = as.Date(sprintf("%d-%02d-15", yr, (qn - 1L) * 3L + 1L)),
      output_type = "point", output_id = NA_character_,
      value_native = agg$value, unit_native = "yoy_pct",
      declared_target = "unspecified",
      retrieved_at = retrieved(f), source_file = paste0(f, ":", name)
    )
  })
  rbindlist(Filter(Negate(is.null), out))
}

# --- assemble ---

main <- function() {
  dir.create(OUT, recursive = TRUE, showWarnings = FALSE)
  gd <- gdpnow()
  gd_fc <- gd$fc
  gd_cur <- gdpnow_current()
  if (nrow(gd_cur)) {
    gd_fc <- rbind(gd_fc, gd_cur)
    gd_fc <- gd_fc[!duplicated(gd_fc, by = c("target_period", "forecast_date"))]
  }
  forecasts <- rbindlist(list(gd_fc, nyfed(), spf_philly(), spf_ecb()))
  forecasts[, value_qq := ifelse(unit_native == "saar_pct",
                                 round(saar_to_qq(value_native), 4), NA_real_)]
  outcomes <- gd$oc
  outcomes[, value_qq := round(saar_to_qq(value_native), 4)]

  setcolorder(forecasts, c("source", "region", "variable", "target_period", "forecast_date",
                           "output_type", "output_id", "value_native", "unit_native", "value_qq",
                           "declared_target", "retrieved_at", "source_file"))
  setcolorder(outcomes, c("region", "variable", "target_period", "release_label", "published_on",
                          "value_native", "unit_native", "value_qq", "source", "retrieved_at",
                          "source_file"))
  setorder(forecasts, source, target_period, forecast_date)
  setorder(outcomes, target_period)

  for (name in c("forecasts", "outcomes")) {
    df <- get(name)
    nanoparquet::write_parquet(df, file.path(OUT, paste0(name, ".parquet")))
    fwrite(df, file.path(OUT, paste0(name, ".csv")))
    cat(sprintf("%s: %s rows -> data/archive/%s.parquet (+.csv)\n",
                name, format(nrow(df), big.mark = ","), name))
  }
  cat("\nforecast rows per source:\n")
  print(forecasts[, .(rows = .N, first = min(forecast_date), last = max(forecast_date)), by = source])
}

main()
